packages/flutter/lib/src/services/text_editing_value.dart

```diff
@@ -0,0 +1,444 @@
+  /// Returns a new TextEditingValue representing the state after the deletion.
+  ///
+  /// If the selection is not collapsed, deletes the selection regardless of the
+  /// given index.
+  TextEditingValue deleteTo(int index, [bool includeWhitespace = true]) {
+    assert(selection != null);
+
+    if (!selection.isValid) {
+      return this;
+    }
+
+    if (!selection.isCollapsed) {
+      return _deleteNonEmptySelection();
+    }
+
+    final String textBefore = selection.textBefore(text);
+    if (index == selection.extentOffset) {
+      return this;
+    } else if (index < selection.extentOffset) {
+      if (textBefore.isEmpty) {
+        return this;
+      }
+      final TextRange nextComposingRange = !composing.isValid || composing.isCollapsed
+        ? TextRange.empty
+        : TextRange(
+          start: composing.start - (composing.start - index).clamp(0, textBefore.length - index),
+          end: composing.end - (composing.end - index).clamp(0, textBefore.length - index),
+        );
+      return TextEditingValue(
+        text: text.substring(0, index) + text.substring(selection.extentOffset, text.length),
+        selection: TextSelection.collapsed(offset: index, affinity: selection.affinity),
+        composing: nextComposingRange,
+      );
+    }
+
+    if (selection.textAfter(text).isEmpty) {
+      return this;
+    }
+    final String nextText = text.substring(0, selection.extentOffset) + text.substring(index, text.length);
+    final int charactersDeleted = text.length - nextText.length;
+    final int selectionToComposingStart = composing.start - selection.baseOffset;
+    final int charactersDeletedBeforeComposingStart = selectionToComposingStart.clamp(0, charactersDeleted);
+    final int selectionToComposingEnd = composing.end - selection.baseOffset;
+    final int charactersDeletedBeforeComposingEnd = selectionToComposingEnd.clamp(0, charactersDeleted);
+    final TextRange nextComposingRange = !composing.isValid || composing.isCollapsed
+      ? TextRange.empty
+      : TextRange(
+        start: composing.start - charactersDeletedBeforeComposingStart,
+        end: composing.end - charactersDeletedBeforeComposingEnd,
+      );
+    return TextEditingValue(
+      text: nextText,
+      selection: TextSelection.collapsed(offset: index - text.length + nextText.length),
+      composing: nextComposingRange,
+    );
+  }
+
+  /// Return [selection] collapsed and moved to the given [TextPosition].
+  TextSelection moveSelectionTo(TextPosition position) {
+    assert(selection != null);
+
+    // If the selection is collapsed at the position already, then nothing
+    // happens.
+    if (selection.isCollapsed && selection.extentOffset == position.offset) {
+      return selection;
+    }
+
+    return selection.copyWith(
+      baseOffset: position.offset,
+      extentOffset: position.offset,
+      affinity: position.affinity,
+    );
+  }
+
+  /// Return [selection] expanded to the given [TextPosition].
```

Comment:
It's not clear to me what the differences are between expand and extend from reading the documentation.

Reply:
I knew that was going to be confusing.  I'll add some docs about it in both expandSelectionTo and extendSelectionTo.

Basically, "extend" moves the extentOffset while holding the baseOffset fixed, but "expand" grows the selection to include the given position.  If you try cmd+shift+left/right on a Mac, that's expand, while option+shift+up/down is extend.

Reply:
I've improved the docs: https://github.com/flutter/flutter/pull/86736/commits/0675ec4ad718f1202101edf3d0c1d6167a4bd9ce

Would it be more intuitive if I renamed "extend" to "pivot"?